File: src/ui/diff.py
"""Pure diff computation for the review screen.

Extracted from src/review.py per docs/UI-REDESIGN.md so the rendered diff views
(diff vs HEAD, diff vs previous draft) can be unit-tested without touching the
Rich-based renderer.
"""
from __future__ import annotations

import difflib

NO_HEAD_DIFF = "(no diff — first draft, no prior README)"
NO_PREV_DIFF = "(no diff — this is the first draft)"
NO_CHANGES = "(no changes)"

_DRAFT_LABEL = "README.md (draft)"
_HEAD_LABEL = "README.md (HEAD)"
_PREV_LABEL = "README.md (prev draft)"


def unified(old: str, new: str, *, fromfile: str, tofile: str) -> str:
    """Return a unified diff between *old* and *new*, or NO_CHANGES sentinel."""
    old_lines = old.splitlines(keepends=True)
    new_lines = new.splitlines(keepends=True)
    out = "".join(
        difflib.unified_diff(old_lines, new_lines, fromfile=fromfile, tofile=tofile)
    )
    return out or NO_CHANGES


def diff_vs_head(head: str | None, current: str) -> str:
    """Diff committed README against current draft. Fallback if no HEAD README."""
    if head is None:
        return NO_HEAD_DIFF
    return unified(head, current, fromfile=_HEAD_LABEL, tofile=_DRAFT_LABEL)


def diff_vs_prev(prev: str | None, current: str) -> str:
    """Diff previous Claude draft against current. Fallback on first iteration."""
    if prev is None:
        return NO_PREV_DIFF
    return unified(prev, current, fromfile=_PREV_LABEL, tofile=_DRAFT_LABEL)

File: go/internal/diff/diff.go
// Package diff renders plain-text diffs of README.md.
package diff

import (
	"context"
	"fmt"
	"os/exec"
	"strings"

	"github.com/hexops/gotextdiff"
	"github.com/hexops/gotextdiff/myers"
	"github.com/hexops/gotextdiff/span"
)

// Sentinels for edge cases.
const (
	NoHeadDiff = "(no diff — first draft, no prior README)"
	NoPrevDiff = "(no diff — this is the first draft)"
	NoChanges  = "(no changes)"
)

const (
	draftLabel = "README.md (draft)"
	headLabel  = "README.md (HEAD)"
	prevLabel  = "README.md (prev draft)"
)

// Unified returns a unified diff between old and new, or NoChanges sentinel.
func Unified(old, new, fromfile, tofile string) string {
	if old == new {
		return NoChanges
	}
	edits := myers.ComputeEdits(span.URIFromPath(fromfile), old, new)
	diff := fmt.Sprint(gotextdiff.ToUnified(fromfile, tofile, old, edits))
	if diff == "" {
		return NoChanges
	}
	return strings.TrimRight(diff, "\n")
}

// DiffVsHead diffs committed README against current draft. Falls back if no HEAD README.
func DiffVsHead(head *string, current string) string {
	if head == nil {
		return NoHeadDiff
	}
	return Unified(*head, current, headLabel, draftLabel)
}

// DiffVsPrev diffs previous Claude draft against current. Falls back on first iteration.
func DiffVsPrev(prev *string, current string) string {
	if prev == nil {
		return NoPrevDiff
	}
	return Unified(*prev, current, prevLabel, draftLabel)
}

// Plain returns `git diff --no-color README.md` from repoDir.
func Plain(ctx context.Context, repoDir string) (string, error) {
	cmd := exec.CommandContext(ctx, "git", "diff", "--no-color", "README.md")
	cmd.Dir = repoDir
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("git diff: %w", err)
	}
	return string(out), nil
}
